File: t-tasks/src/main/java/com/teo/ttasks/ui/fragments/task_lists/TaskListsPresenter.java
package com.teo.ttasks.ui.fragments.task_lists;

import android.support.annotation.NonNull;

import com.teo.ttasks.data.model.TTaskList;
import com.teo.ttasks.data.local.TaskListFields;
import com.teo.ttasks.data.remote.TasksHelper;
import com.teo.ttasks.ui.base.Presenter;
import com.teo.ttasks.ui.items.TaskListItem;

import java.util.ArrayList;
import java.util.List;

import io.realm.Realm;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import timber.log.Timber;

public class TaskListsPresenter extends Presenter<TaskListsView> {

    private final TasksHelper tasksHelper;

    private TaskListFields taskListFields;

    private Realm realm;

    public TaskListsPresenter(TasksHelper tasksHelper) {
        this.tasksHelper = tasksHelper;
        taskListFields = new TaskListFields();
    }

    /**
     * Load all the user's task lists.
     */
    void getTaskLists() {
        {
            final TaskListsView view = view();
            if (view != null) view.onTaskListsLoading();
        }
        final Subscription subscription = tasksHelper.getTaskLists(realm)
                .map(taskLists -> {
                    List<TaskListItem> taskListItems = new ArrayList<>(taskLists.size());

                    for (int i = 0, taskListsSize = taskLists.size(); i < taskListsSize; i++) {
                        TTaskList taskList = taskLists.get(i);
                        taskListItems.add(new TaskListItem(taskList, tasksHelper.getTaskListSize(taskList.getId(), realm)));
                    }

                    return taskListItems;
                })
                .subscribe(taskListItems -> {
                    final TaskListsView view = view();
                    if (view != null) {
                        if (taskListItems == null) view.onTaskListsError();
                        else if (taskListItems.isEmpty()) view.onTaskListsEmpty();
                        else view.onTaskListsLoaded(taskListItems);
                    }
                });
        unsubscribeOnUnbindView(subscription);
    }

    /**
     * Set the task list title.
     *
     * @param title the new task list title
     */
    void setTaskListTitle(String title) {
        taskListFields.putTitle(title);
    }

    void createTaskList() {
        // Nothing entered
        if (taskListFields.isEmpty())
            return;

        // Create the task list offline
        TTaskList tTaskList = TasksHelper.createTaskList(taskListFields);
        Timber.d("New task list with id %s", tTaskList.getId());
        realm.executeTransaction(realm -> realm.insertOrUpdate(tTaskList));

        // Create the task remotely
        tasksHelper.newTaskList(taskListFields)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        taskList -> {
                            // Update the local task with the full information and delete the old task
                            TTaskList managedTaskList = tasksHelper.getTaskList(tTaskList.getId(), realm);
                            realm.executeTransaction(realm -> {
                                managedTaskList.getTaskList().deleteFromRealm();
                                managedTaskList.switchTaskList(realm.copyToRealm(taskList));
                                managedTaskList.setSynced(true);
                            });
                            Timber.d("Task list id updated to %s", managedTaskList.getId());
                        },
                        throwable -> {
                            Timber.e(throwable.toString());
                        }
                );
    }

    /**
     * Update the task list with the specified ID. Currently, only a title change is supported.
     *
     * @param taskListId task list identifier
     * @param isOnline   flag indicating an active network connection
     */
    void updateTaskList(String taskListId, boolean isOnline) {
        // Nothing changed
        if (taskListFields.isEmpty())
            return;

        // Update locally
        final TTaskList managedTaskList = tasksHelper.getTaskList(taskListId, realm);
        realm.executeTransaction(realm -> {
            managedTaskList.update(taskListFields);
            managedTaskList.setSynced(false);
        });

        // Update remotely
        if (isOnline) {
            tasksHelper.updateTaskList(taskListId, taskListFields)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            taskList -> {
                                realm.executeTransaction(realm -> {
                                    realm.insertOrUpdate(taskList);
                                    managedTaskList.setSynced(true);
                                });
                            },
                            throwable -> {
                                Timber.e(throwable.toString());
                                final TaskListsView view = view();
                                if (view != null) view.onTaskListUpdateError();
                            }
                    );
        }
    }

    /**
     * Delete the task list with the specified ID.
     *
     * @param taskListId task list identifier
     */
    void deleteTaskList(String taskListId) {
        // Get the task list
        final TTaskList managedTaskList = tasksHelper.getTaskList(taskListId, realm);

        // Mark it as deleted so it doesn't show up in the list
        realm.executeTransaction(realm -> managedTaskList.setDeleted(true));

        // Delete the task list
        tasksHelper.deleteTaskList(taskListId)
                .subscribe(
                        aVoid -> { /* Do nothing */ },
                        throwable -> {
                            Timber.e(throwable.toString());
                        }
                );
    }

    @Override
    public void bindView(@NonNull TaskListsView view) {
        super.bindView(view);
        realm = Realm.getDefaultInstance();
    }

    @Override
    public void unbindView(@NonNull TaskListsView view) {
        super.unbindView(view);
        realm.close();
    }
}

File: t-tasks/src/main/java/com/teo/ttasks/ui/fragments/task_lists/TaskListsPresenter.kt
package com.teo.ttasks.ui.fragments.task_lists

import com.teo.ttasks.data.local.TaskListFields
import com.teo.ttasks.data.remote.TasksHelper
import com.teo.ttasks.ui.base.Presenter
import com.teo.ttasks.ui.items.TaskListItem
import io.reactivex.android.schedulers.AndroidSchedulers
import io.realm.Realm
import timber.log.Timber
import java.util.*

internal class TaskListsPresenter(private val tasksHelper: TasksHelper) : Presenter<TaskListsView>() {

    private val taskListFields: TaskListFields = TaskListFields()

    private lateinit var realm: Realm

    /**
     * Load all the user's task lists.
     */
    internal fun getTaskLists() {
        view()?.onTaskListsLoading()

        val disposable = tasksHelper.getTaskLists(realm)
                .map { taskLists ->
                    val taskListItems = ArrayList<TaskListItem>(taskLists.size)

                    taskLists.forEach {
                        taskListItems.add(TaskListItem(it, tasksHelper.getTaskListSize(it.id, realm)))
                    }

                    taskListItems
                }
                .subscribe { taskListItems ->
                    when {
                        taskListItems == null -> view()?.onTaskListsError()
                        taskListItems.isEmpty() -> view()?.onTaskListsEmpty()
                        else -> view()?.onTaskListsLoaded(taskListItems)
                    }
                }
        disposeOnUnbindView(disposable)
    }

    /**
     * Set the task list title.

     * @param title the new task list title
     */
    internal fun setTaskListTitle(title: String) {
        taskListFields.title = title
    }

    internal fun createTaskList() {
        // Nothing entered
        if (taskListFields.isEmpty())
            return

        // Create the task list offline
        val localTaskList = taskListFields.toTaskList()
        Timber.d("New task list with id %s", localTaskList.id)
        realm.executeTransaction { it.insertOrUpdate(localTaskList) }

        // Create the task remotely
        // TODO schedule job for this
        tasksHelper.createTaskList(taskListFields)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { taskList ->
                            // Delete the old task list and insert the new one
                            val managedTaskList = tasksHelper.getTaskList(localTaskList.id, realm)!!
                            realm.executeTransaction {
                                managedTaskList.deleteFromRealm()
                                taskList.synced = true
                                it.insertOrUpdate(taskList)
                            }
                            Timber.d("Task list id updated to %s", managedTaskList.id)
                        },
                        { Timber.e(it, "Error while creating a new task list") }
                )
    }

    /**
     * Update the task list with the specified ID. Currently, only a title change is supported.

     * @param taskListId task list identifier
     * *
     * @param isOnline   flag indicating an active network connection
     */
    internal fun updateTaskList(taskListId: String, isOnline: Boolean) {
        // Nothing changed
        if (taskListFields.isEmpty())
            return

        // Update locally
        val managedTaskList = tasksHelper.getTaskList(taskListId, realm)!!
        realm.executeTransaction {
            managedTaskList.update(taskListFields)
            managedTaskList.synced = false
        }

        // Update remotely
        if (isOnline) {
            tasksHelper.updateTaskList(taskListId, taskListFields)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            { taskList ->
                                realm.executeTransaction {
                                    it.insertOrUpdate(taskList)
                                    managedTaskList.synced = true
                                }
                            },
                            {
                                Timber.e(it, "Error while updating task list %s", taskListId)
                                view()?.onTaskListUpdateError()
                            }
                    )
        }
    }

    /**
     * Delete the task list with the specified ID.

     * @param taskListId task list identifier
     */
    internal fun deleteTaskList(taskListId: String) {
        // Get the task list
        val managedTaskList = tasksHelper.getTaskList(taskListId, realm)!!

        // Mark it as deleted so it doesn't show up in the list
        realm.executeTransaction { managedTaskList.deleted = true }

        // Delete the task list
        tasksHelper.deleteTaskList(taskListId)
                .subscribe({ /* Do nothing */ }, { throwable -> Timber.e(throwable.toString()) })
    }

    override fun bindView(view: TaskListsView) {
        super.bindView(view)
        realm = Realm.getDefaultInstance()
    }

    override fun unbindView(view: TaskListsView) {
        super.unbindView(view)
        realm.close()
    }
}
